master/buildbot/process/builder.py

```diff
@@ -319,7 +320,7 @@ def run_cleanups():
 
         # set up locks
         build.setLocks(self.config.locks)
-        cleanups.append(lambda: slavebuilder.slave.releaseLocks())
+        cleanups.append(lambda: slavebuilder.slave.releaseLocks() if slavebuilder.slave is not None else None)
```

Comment:
I'm not a fan of this.  This implies that there is something racy here that we don't really understand, and we're just hacking around the problem.

The real problem here would seem to be that build.setLocks() and slavebuiler.slave.releaseLocks() are not really the complementary operations that the code is assuming.  _That_ is the thing that really needs to be fixed.

Also, do you have an upstream bug that you can reference?